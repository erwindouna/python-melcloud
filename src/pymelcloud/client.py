"""MEL API access."""

from datetime import UTC, datetime, timedelta
from typing import Any

from aiohttp import ClientSession

BASE_URL = "https://app.melcloud.com/Mitsubishi.Wifi.Client"


def _headers(token: str) -> dict[str, str]:
    return {
        "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:73.0) "
        "Gecko/20100101 Firefox/73.0",
        "Accept": "application/json, text/javascript, */*; q=0.01",
        "Accept-Language": "en-US,en;q=0.5",
        "Accept-Encoding": "gzip, deflate, br",
        "X-MitsContextKey": token,
        "X-Requested-With": "XMLHttpRequest",
        "Cookie": "policyaccepted=true",
    }


async def _do_login(
    _session: ClientSession, email: str, password: str
) -> dict[str, Any]:
    body = {
        "Email": email,
        "Password": password,
        "Language": 0,
        "AppVersion": "1.19.1.1",
        "Persist": True,
        "CaptchaResponse": None,
    }

    async with _session.post(
        f"{BASE_URL}/Login/ClientLogin", json=body, raise_for_status=True
    ) as resp:
        result: dict[str, Any] = await resp.json()
        return result


async def login(
    email: str,
    password: str,
    session: ClientSession | None = None,
    *,
    user_update_interval: timedelta | None = None,
    conf_update_interval: timedelta | None = None,
    device_set_debounce: timedelta | None = None,
) -> "Client":
    """Login using email and password."""
    if session:
        response = await _do_login(session, email, password)
    else:
        async with ClientSession() as _session:
            response = await _do_login(_session, email, password)

    login_data = response.get("LoginData")
    if login_data is None:
        raise ValueError("No login data in response")

    context_key = login_data.get("ContextKey")
    if context_key is None:
        raise ValueError("No context key in login data")

    return Client(
        context_key,
        session,
        user_update_interval=user_update_interval or timedelta(minutes=5),
        conf_update_interval=conf_update_interval or timedelta(seconds=59),
        device_set_debounce=device_set_debounce or timedelta(seconds=1),
    )


class Client:
    """MELCloud client.

    Please do not use this class directly. It is better to use the get_devices
    method exposed by the __init__.py.
    """

    def __init__(
        self,
        token: str,
        session: ClientSession | None = None,
        *,
        user_update_interval: timedelta = timedelta(minutes=5),
        conf_update_interval: timedelta = timedelta(seconds=59),
        device_set_debounce: timedelta = timedelta(seconds=1),
    ) -> None:
        """Initialize MELCloud client."""
        self._token = token
        if session:
            self._session = session
            self._managed_session = False
        else:
            self._session = ClientSession()
            self._managed_session = True
        self._user_update_interval = user_update_interval
        self._conf_update_interval = conf_update_interval
        self._device_set_debounce = device_set_debounce

        self._last_user_update: datetime | None = None
        self._last_conf_update: datetime | None = None
        self._device_confs: list[dict[str, Any]] = []
        self._account: dict[str, Any] | None = None

    @property
    def token(self) -> str:
        """Return currently used token."""
        return self._token

    @property
    def device_confs(self) -> list[dict[Any, Any]]:
        """Return device configurations."""
        return self._device_confs

    @property
    def account(self) -> dict[Any, Any] | None:
        """Return account."""
        return self._account

    async def _fetch_user_details(self) -> None:
        """Fetch user details."""
        async with self._session.get(
            f"{BASE_URL}/User/GetUserDetails",
            headers=_headers(self._token),
            raise_for_status=True,
        ) as resp:
            self._account = await resp.json()

    async def _fetch_device_confs(self) -> None:
        """Fetch all configured devices."""
        url = f"{BASE_URL}/User/ListDevices"
        async with self._session.get(
            url, headers=_headers(self._token), raise_for_status=True
        ) as resp:
            entries = await resp.json()
            new_devices: list[dict[str, Any]] = []
            for entry in entries:
                new_devices = new_devices + entry["Structure"]["Devices"]

                for area in entry["Structure"]["Areas"]:
                    new_devices = new_devices + area["Devices"]

                for floor in entry["Structure"]["Floors"]:
                    new_devices = new_devices + floor["Devices"]

                    for area in floor["Areas"]:
                        new_devices = new_devices + area["Devices"]

            visited: set[Any] = set()
            filtered_devices = []
            for d in new_devices:
                device_id = d["DeviceID"]
                if device_id not in visited:
                    visited.add(device_id)
                    filtered_devices.append(d)
            self._device_confs = filtered_devices

    async def update_confs(self) -> None:
        """Update device_confs and account.

        Calls are rate limited to allow Device instances to freely poll their own
        state while refreshing the device_confs list and account.
        """
        now = datetime.now(tz=UTC)

        if (
            self._last_conf_update is None
            or now - self._last_conf_update > self._conf_update_interval
        ):
            await self._fetch_device_confs()
            self._last_conf_update = now

        if (
            self._last_user_update is None
            or now - self._last_user_update > self._user_update_interval
        ):
            await self._fetch_user_details()
            self._last_user_update = now

    async def fetch_device_units(self, device: Any) -> dict[Any, Any] | None:
        """Fetch unit information for a device.

        User provided info such as indoor/outdoor unit model names and
        serial numbers.
        """
        async with self._session.post(
            f"{BASE_URL}/Device/ListDeviceUnits",
            headers=_headers(self._token),
            json={"deviceId": device.device_id},
            raise_for_status=True,
        ) as resp:
            result: dict[Any, Any] = await resp.json()
            return result

    async def fetch_device_state(self, device: Any) -> dict[Any, Any] | None:
        """Fetch state information of a device.

        This method should not be called more than once a minute. Rate
        limiting is left to the caller.
        """
        device_id = device.device_id
        building_id = device.building_id
        async with self._session.get(
            f"{BASE_URL}/Device/Get?id={device_id}&buildingID={building_id}",
            headers=_headers(self._token),
            raise_for_status=True,
        ) as resp:
            result: dict[Any, Any] = await resp.json()
            return result

    async def fetch_energy_report(self, device: Any) -> dict[Any, Any] | None:
        """Fetch energy report containing today and 1-2 days from the past."""
        device_id = device.device_id
        now = datetime.now(tz=UTC)
        from_str = (now - timedelta(days=2)).strftime("%Y-%m-%d")
        to_str = (now + timedelta(days=2)).strftime("%Y-%m-%d")

        async with self._session.post(
            f"{BASE_URL}/EnergyCost/Report",
            headers=_headers(self._token),
            json={
                "DeviceId": device_id,
                "UseCurrency": False,
                "FromDate": f"{from_str}T00:00:00",
                "ToDate": f"{to_str}T00:00:00",
            },
            raise_for_status=True,
        ) as resp:
            result: dict[Any, Any] = await resp.json()
            return result

    async def set_device_state(self, device: Any) -> None:
        """Update device state.

        This method is as dumb as it gets. Device is responsible for updating
        the state and managing EffectiveFlags.
        """
        device_type = device.get("DeviceType")
        if device_type == 0:
            setter = "SetAta"
        elif device_type == 1:
            setter = "SetAtw"
        elif device_type == 3:
            setter = "SetErv"
        else:
            raise ValueError(f"Unsupported device type [{device_type}]")

        async with self._session.post(
            f"{BASE_URL}/Device/{setter}",
            headers=_headers(self._token),
            json=device,
            raise_for_status=True,
        ) as resp:
            # Read response but don't return it since method returns None
            await resp.json()
