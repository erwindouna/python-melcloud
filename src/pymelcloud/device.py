"""Base MELCloud device."""

import asyncio
from abc import ABC, abstractmethod
from datetime import UTC, datetime, timedelta
from decimal import ROUND_HALF_UP, Decimal
from typing import Any

from pymelcloud.client import Client
from pymelcloud.const import (
    ACCESS_LEVEL,
    DEVICE_TYPE_LOOKUP,
    DEVICE_TYPE_UNKNOWN,
    UNIT_TEMP_CELSIUS,
    UNIT_TEMP_FAHRENHEIT,
)

PROPERTY_POWER = "power"

EFFECTIVE_FLAGS = "EffectiveFlags"
HAS_PENDING_COMMAND = "HasPendingCommand"


class Device(ABC):
    """MELCloud base device representation."""

    def __init__(
        self,
        device_conf: dict[str, Any],
        client: Client,
        set_debounce: timedelta = timedelta(seconds=1),
    ) -> None:
        """Initialize a device."""
        self.device_id = device_conf.get("DeviceID")
        self.building_id = device_conf.get("BuildingID")
        self.mac = device_conf.get("MacAddress")
        self.serial = device_conf.get("SerialNumber")
        self.access_level = device_conf.get("AccessLevel")

        self._use_fahrenheit = False
        if client.account is not None:
            self._use_fahrenheit = client.account.get("UseFahrenheit", False)

        self._device_conf = device_conf
        self._state: dict[str, Any] | None = None
        self._device_units: dict[str, Any] | None = None
        self._energy_report: dict[str, Any] | None = None
        self._client = client

        self._set_debounce = set_debounce
        self._set_event = asyncio.Event()
        self._write_task: asyncio.Future[None] | None = None
        self._pending_writes: dict[str, Any] = {}

    def get_device_prop(self, name: str) -> Any | None:
        """Access device properties while shortcutting the nested device access."""
        device = self._device_conf.get("Device", {})
        return device.get(name)

    def get_state_prop(self, name: str) -> Any | None:
        """Access state prop without None check."""
        if self._state is None:
            return None
        return self._state.get(name)

    def round_temperature(self, temperature: float) -> float:
        """Round a temperature to the nearest temperature increment."""
        return (
            float(
                Decimal(str(temperature / self.temperature_increment)).quantize(
                    Decimal("1"), rounding=ROUND_HALF_UP
                )
            )
            * self.temperature_increment
        )

    @abstractmethod
    def apply_write(self, state: dict[str, Any], key: str, value: Any) -> None:
        """Apply writes to state object.

        Used for property validation, do not modify device state.
        """

    async def update(self) -> None:
        """Fetch state of the device from MELCloud.

        List of device_confs is also updated.

        Please, rate limit calls to this method. Polling every 60 seconds should be
        enough to catch all events at the rate they are coming in to MELCloud with the
        exception of changes performed through MELCloud directly.
        """
        await self._client.update_confs()
        self._device_conf = next(
            c
            for c in self._client.device_confs
            if c.get("DeviceID") == self.device_id
            and c.get("BuildingID") == self.building_id
        )
        self._state = await self._client.fetch_device_state(self)
        self._energy_report = await self._client.fetch_energy_report(self)

        if self._device_units is None and self.access_level != ACCESS_LEVEL.get(
            "GUEST"
        ):
            self._device_units = await self._client.fetch_device_units(self)

    async def set(self, properties: dict[str, Any]) -> None:
        """Schedule property write to MELCloud."""
        if self._write_task is not None:
            self._write_task.cancel()

        for k, value in properties.items():
            if k == PROPERTY_POWER:
                continue
            self.apply_write({}, k, value)

        self._pending_writes.update(properties)

        self._write_task = asyncio.ensure_future(self._write())
        await self._set_event.wait()

    async def _write(self) -> None:
        await asyncio.sleep(self._set_debounce.total_seconds())
        if self._state is None:
            return
        new_state = self._state.copy()

        for k, value in self._pending_writes.items():
            if k == PROPERTY_POWER:
                new_state["Power"] = value
                new_state[EFFECTIVE_FLAGS] = new_state.get(EFFECTIVE_FLAGS, 0) | 0x01
            else:
                self.apply_write(new_state, k, value)

        if new_state[EFFECTIVE_FLAGS] != 0:
            new_state.update({HAS_PENDING_COMMAND: True})

        self._pending_writes = {}
        await self._client.set_device_state(new_state)
        self._state = new_state  # Update our local state with the changes we sent
        self._set_event.set()
        self._set_event.clear()

    @property
    def name(self) -> str:
        """Return device name."""
        name = self._device_conf.get("DeviceName", "")
        return str(name) if name is not None else ""

    @property
    def device_type(self) -> str:
        """Return type of the device."""
        return DEVICE_TYPE_LOOKUP.get(
            self._device_conf.get("Device", {}).get("DeviceType", -1),
            DEVICE_TYPE_UNKNOWN,
        )

    @property
    def units(self) -> list[dict[str, Any]] | None:
        """Return device model info."""
        if self._device_units is None:
            return None

        # _device_units might be a dict containing a list or a list directly
        units_data = self._device_units
        if isinstance(units_data, dict):
            # If it's a dict, it might contain the units in a key like "Units"
            units_list = units_data.get("Units", [])
        elif isinstance(units_data, list):
            units_list = units_data
        else:
            return []

        return [
            {
                "model_number": unit.get("ModelNumber"),
                "model": unit.get("Model"),
                "serial_number": unit.get("SerialNumber"),
            }
            for unit in units_list
            if isinstance(unit, dict)
        ]

    @property
    def temp_unit(self) -> str:
        """Return temperature unit used by the device."""
        if self._use_fahrenheit:
            return UNIT_TEMP_FAHRENHEIT
        return UNIT_TEMP_CELSIUS

    @property
    def temperature_increment(self) -> float:
        """Return temperature increment."""
        increment = self._device_conf.get("Device", {}).get("TemperatureIncrement", 0.5)
        return float(increment) if increment is not None else 0.5

    @property
    def last_seen(self) -> datetime | None:
        """Return timestamp of the last communication from device to MELCloud.

        The timestamp is in UTC.
        """
        if self._state is None:
            return None
        last_communication = self._state.get("LastCommunication")
        if not isinstance(last_communication, str):
            return None
        return datetime.strptime(last_communication, "%Y-%m-%dT%H:%M:%S.%f").replace(
            tzinfo=UTC
        )

    @property
    def power(self) -> bool | None:
        """Return power on / standby state of the device."""
        if self._state is None:
            return None
        return self._state.get("Power")

    @property
    def daily_energy_consumed(self) -> float | None:
        """Return daily energy consumption for the current day in kWh.

        The value resets at midnight MELCloud time. The logic here is a bit iffy and
        fragmented between Device and Client. Here's how it goes:
          - Client requests a 5 day report. Today, 2 days from the past and 2 days from
            the past.
          - MELCloud, with its clock potentially set to a different timezone than the
            client, returns a report containing data from a couple of days from the
            past and from the current day in MELCloud time.
          - Device sums up the date from the last day bucket in the report.

        TLDR: Request some days from the past and some days from the future -> receive
        the latest day bucket.
        """
        if self._energy_report is None:
            return None

        consumption: float = 0.0

        for mode in ["Heating", "Cooling", "Auto", "Dry", "Fan", "Other"]:
            previous_reports = self._energy_report.get(mode, [0.0])
            last_report = previous_reports[-1] if previous_reports else 0.0
            consumption += float(last_report)

        return consumption

    @property
    def wifi_signal(self) -> int | None:
        """Return wifi signal in dBm (negative value)."""
        if self._device_conf is None:
            return None
        signal = self._device_conf.get("Device", {}).get("WifiSignalStrength", None)
        return (
            int(signal)
            if signal is not None and isinstance(signal, (int, float))
            else None
        )

    @property
    def has_error(self) -> bool:
        """Return True if the device has error state."""
        if self._state is None:
            return False
        has_error = self._state.get("HasError", False)
        return bool(has_error)

    @property
    def error_code(self) -> int | None:
        """Return error code.

        This is a property that probably should be checked if "has_error" = true.
        Till now I have a fixed code = 8000 and never have error on the units.
        """
        if self._state is None:
            return None
        error_code = self._state.get("ErrorCode", None)
        return int(error_code) if error_code is not None else None
