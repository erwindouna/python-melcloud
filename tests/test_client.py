"""Tests for the client module."""

from __future__ import annotations

import aiohttp
import pytest
from aioresponses import aioresponses
from syrupy import SnapshotAssertion

import pymelcloud
from pymelcloud.client import BASE_URL, Client
from tests import load_fixture


async def test_create_session(responses: aioresponses) -> None:
    """Test creating a session."""
    responses.post(
        f"{BASE_URL}/Login/ClientLogin", status=200, body=load_fixture("login.json")
    )
    async with aiohttp.ClientSession():
        token = await pymelcloud.login(email="username", password="password")
        assert token is not None

    # Test creating a session with a custom session
    responses.post(
        f"{BASE_URL}/Login/ClientLogin", status=200, body=load_fixture("login.json")
    )
    async with aiohttp.ClientSession():
        token = await pymelcloud.login(
            email="username", password="password", session=aiohttp.ClientSession()
        )
        assert token is not None


async def test_login_success(
    responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test logging in with valid credentials."""
    responses.post(
        f"{BASE_URL}/Login/ClientLogin", status=200, body=load_fixture("login.json")
    )
    async with aiohttp.ClientSession() as session:
        client = await pymelcloud.client.login(
            email="username", password="password", session=session
        )

        # pylint: disable=protected-access
        assert client is not None
        assert isinstance(client, Client)
        assert client.token == "0-XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
        assert client._session == session
        assert client._managed_session is False
        assert client == snapshot


@pytest.mark.skip(
    reason="Skipping, can be implemented later when exceptions are in place."
)
async def test_login_failure(responses: aioresponses) -> None:
    """Test logging in with invalid credentials."""
    responses.post(
        f"{BASE_URL}/Login/ClientLogin",
        status=200,
        body=load_fixture("login_failure.json"),
    )
    async with aiohttp.ClientSession() as session:
        with pytest.raises(aiohttp.ClientResponseError):
            await pymelcloud.client.login(
                email="username", password="password", session=session
            )


async def test_update_confs(
    melcloud_client: Client, responses: aioresponses, snapshot: SnapshotAssertion
) -> None:
    """Test updating device configurations."""
    responses.get(
        f"{BASE_URL}/User/ListDevices",
        status=200,
        body=load_fixture("list_devices.json"),
    )
    responses.get(
        f"{BASE_URL}/User/GetUserDetails",
        status=200,
        body=load_fixture("user_details.json"),
    )

    # pylint: disable=protected-access
    await melcloud_client.update_confs()
    assert melcloud_client._device_confs == snapshot
    assert melcloud_client._account == snapshot
