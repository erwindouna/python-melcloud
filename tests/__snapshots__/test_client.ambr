# serializer version: 1
# name: test_login_success
  Client(
    account=None,
    device_confs=list([
    ]),
    token='0-XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX',
  )
# ---
# name: test_update_confs
  list([
    dict({
      'AccessLevel': 3,
      'AdaptorType': 3,
      'AreaID': None,
      'AreaName': None,
      'BuildingCountry': None,
      'BuildingID': 1,
      'BuildingName': None,
      'Device': dict({
        'ActualFanSpeed': 0,
        'AdaptorType': 3,
        'AirDirectionFunction': True,
        'AutoEnergyConsumedRate1': 0,
        'AutoEnergyConsumedRate2': 0,
        'AutomaticFanSpeed': False,
        'CanAuto': False,
        'CanCool': True,
        'CanDry': True,
        'CanHeat': True,
        'ConfiguredDemandPercentage': None,
        'ConsecutiveUploadErrors': 0,
        'CoolingDisabled': False,
        'CoolingEnergyConsumedRate1': 0,
        'CoolingEnergyConsumedRate2': 0,
        'CurrentEnergyConsumed': 0,
        'CurrentEnergyMode': 7,
        'DefaultCoolingSetTemperature': 10.0,
        'DefaultHeatingSetTemperature': None,
        'DemandPercentage': 100,
        'DetectedCountry': None,
        'DeviceID': 1,
        'DeviceType': 0,
        'DiagnosticEndDate': None,
        'DiagnosticMode': 0,
        'DoNotRespondAfter': None,
        'DryEnergyConsumedRate1': 0,
        'DryEnergyConsumedRate2': 0,
        'EffectiveFlags': 0,
        'EffectivePCycle': 1,
        'EnergyCorrectionActive': False,
        'EnergyCorrectionModel': None,
        'ErrorCode': 8000,
        'ErrorMessages': '',
        'EstimateAtaEnergyProduction': False,
        'EstimateAtaEnergyProductionOptIn': False,
        'EstimateAtaEnergyProductionOptInTimestamp': None,
        'ExceptionCount': None,
        'ExceptionDate': None,
        'ExceptionHash': None,
        'ExpectedCommand': 1,
        'FanEnergyConsumedRate1': 0,
        'FanEnergyConsumedRate2': 0,
        'FanSpeed': 1,
        'FirmwareAppVersion': 0,
        'FirmwareDeployment': None,
        'FirmwareUpdateAborted': False,
        'FirmwareWebVersion': 0,
        'FirmwareWlanVersion': 0,
        'FlashWrites': 9,
        'HasAutomaticFanSpeed': True,
        'HasDemandSideControl': False,
        'HasEnergyConsumedMeter': True,
        'HasError': False,
        'HasErrorMessages': False,
        'HasHalfDegreeIncrements': True,
        'HasOutdoorTemperature': True,
        'HasZone2': False,
        'HeatingEnergyConsumedRate1': 0,
        'HeatingEnergyConsumedRate2': 0,
        'HideEnergyReport': False,
        'InStandbyMode': False,
        'IsSplitSystem': False,
        'LastEffectiveFlags': 0,
        'LastReset': '2024-04-22T19:00:19.229',
        'LastTimeStamp': '2025-04-08T23:09:00',
        'LegacyDevice': False,
        'LinkToMELCloudHome': False,
        'LinkedByUserFromMELCloudHome': '00000000-0000-0000-0000-000000000000',
        'LinkedDevice': None,
        'MacAddress': '00:00:00:00:00:00',
        'MaxIndoorUnits': 255,
        'MaxOutdoorUnits': 255,
        'MaxPcycle': 1,
        'MaxTempAutomatic': 10.0,
        'MaxTempCoolDry': 10.0,
        'MaxTempHeat': 10.0,
        'MaxTemperatureControlUnits': 0,
        'MinPcycle': 1,
        'MinTempAutomatic': 16.0,
        'MinTempCoolDry': 16.0,
        'MinTempHeat': 10.0,
        'ModelCode': '0000',
        'ModelDisableEnergyReport': False,
        'ModelIsAirCurtain': False,
        'ModelSupportsAuto': True,
        'ModelSupportsDry': True,
        'ModelSupportsEnergyReporting': False,
        'ModelSupportsFanSpeed': True,
        'ModelSupportsHeat': True,
        'ModelSupportsStandbyMode': True,
        'ModelSupportsVaneHorizontal': True,
        'ModelSupportsVaneVertical': True,
        'ModelSupportsWideVane': False,
        'MqttFlags': 0,
        'NumberOfFanSpeeds': 4,
        'Offline': False,
        'OperationMode': 7,
        'OtherEnergyConsumedRate1': 0,
        'OtherEnergyConsumedRate2': 0,
        'OutdoorTemperature': None,
        'Owner': 1,
        'OwnerCountry': 1,
        'OwnerRoleAccessLevel': 1,
        'PCycle': 10,
        'PCycleActual': 0,
        'PCycleConfigured': None,
        'Passcode': None,
        'Position': 'Unknown',
        'Power': False,
        'ProhibitOperationMode': False,
        'ProhibitPower': False,
        'ProhibitSetTemperature': False,
        'ProtocolVersion': 0,
        'Rate1StartTime': None,
        'Rate2StartTime': None,
        'RecordNumMax': 0,
        'RoomTemperature': 15.5,
        'RoomTemperatureLabel': 0,
        'SPTimeout': 1,
        'SSLExpirationDate': '2037-12-31T00:00:00',
        'Scene': None,
        'SerialNumber': '0000000000',
        'ServerCommunicationDisabled': False,
        'SetTemperature': 16.0,
        'SupportsHourlyEnergyReport': True,
        'SwingFunction': True,
        'TemperatureIncrement': 0.5,
        'TemperatureIncrementOverride': 0,
        'TimeZoneID': 1,
        'UnitSupportsStandbyMode': True,
        'UnitVersion': 0,
        'Units': list([
          dict({
            'Device': 0,
            'ID': 1,
            'IsIndoor': True,
            'Model': '1',
            'ModelNumber': None,
            'SerialNumber': None,
            'UnitType': 1,
          }),
        ]),
        'UseTemperatureA': True,
        'VaneHorizontalDirection': 3,
        'VaneHorizontalSwing': False,
        'VaneVerticalDirection': 1,
        'VaneVerticalSwing': False,
        'WeatherForecast': list([
        ]),
        'WifiAdapterStatus': 'NORMAL',
        'WifiSignalStrength': -39,
      }),
      'DeviceID': 1,
      'DeviceName': 'Device1',
      'DirectAccess': False,
      'EndDate': '2500-01-01T00:00:00',
      'EstimateAtaEnergyProduction': False,
      'EstimateAtaEnergyProductionOptIn': False,
      'FloorID': None,
      'FloorName': None,
      'HideDryModeControl': False,
      'HideOutdoorTemperature': False,
      'HideRoomTemperature': False,
      'HideSupplyTemperature': False,
      'HideVaneControls': False,
      'ImageID': -1,
      'InstallationDate': '2023-09-30T00:00:00',
      'LastServiceDate': '2023-09-30T00:00:00',
      'LinkedDevice': None,
      'MacAddress': '00:00:00:00:00:00',
      'MaxTemperature': 10,
      'MinTemperature': 10,
      'OwnerCountry': None,
      'OwnerEmail': None,
      'OwnerID': 1,
      'OwnerName': None,
      'Presets': list([
      ]),
      'SerialNumber': '0000000000',
      'Type': 0,
      'Zone1Name': None,
      'Zone2Name': None,
    }),
  ])
# ---
# name: test_update_confs.1
  dict({
    'Address1': None,
    'Address2': None,
    'AlternateEmailAddress': None,
    'City': None,
    'CountryID': 1,
    'CountryName': 'Country',
    'DatePasswordResetRequested': None,
    'DateVerificationSent': '2023-10-19T15:45:56.55',
    'DateVerified': '2023-10-19T15:46:14.483',
    'DefaultMaximumDevices': 10,
    'EmailAddress': 'user@example.com',
    'EmailCommsErrors': 0,
    'EmailOnCommsError': False,
    'EmailOnUnitError': False,
    'EmailUnitErrors': 0,
    'Enabled': True,
    'FirstName': 'John',
    'HidePresetPanel': False,
    'Honorific': 'Mr.',
    'ID': 1,
    'LanguageID': 0,
    'LanguageName': 'English',
    'LastAnnouncement': 0,
    'LastLogin': None,
    'LastName': 'Doe',
    'MapLatitude': 0.0,
    'MapLongitude': 0.0,
    'MapZoom': 0,
    'MaximumDevices': 0,
    'MobilePhone': None,
    'PendingEmailAddress': None,
    'Phone': None,
    'PostCode': None,
    'RoleID': 0,
    'RoleName': 'Ordinary User',
    'SendMarketingMessages': False,
    'Sex': 'M',
    'SkipFrostProtectionPage1': False,
    'SkipHolidayModePage1': False,
    'SkipTimerPage1': False,
    'TelephoneQuestion': 'Confidential',
    'Terms': 1292,
    'Throttle': True,
    'TimeZoneCity': 1,
    'UseFahrenheit': False,
  })
# ---
